### PR TITLE
lightning: add timeout for "write" RPC

### LINKED ISSUE #46321: lightning stuck in `writeToTiKV` after inject tikv-failure for 60 minutes
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

inject chaos error

```
cases:
  - name: tikv_failure
    faultType: failure
    selector: tikv(random)
    warmUpTime: 1m
    faultDuration: 60m
    recoveryTime: 5m
```

### 3. What did you see instead (Required)

```
goroutine 11915 [select, 421 minutes]:
google.golang.org/grpc/internal/transport.(*http2Client).NewStream(0xc004e638c0, {0x59c7d10, 0xc03ca11f20}, 0xc03e701aa0)
	/go/pkg/mod/google.golang.org/grpc@v1.54.0/internal/transport/http2_client.go:829 +0x5a5
google.golang.org/grpc.(*csAttempt).newStream(0xc03cf4fee0)
	/go/pkg/mod/google.golang.org/grpc@v1.54.0/stream.go:486 +0xb2
google.golang.org/grpc.newClientStreamWithParams.func2(0xc03cf4fee0)
	/go/pkg/mod/google.golang.org/grpc@v1.54.0/stream.go:336 +0x3a
google.golang.org/grpc.(*clientStream).withRetry(0xc03e46bb00, 0xc03e7183e0, 0xc03bb32998)
	/go/pkg/mod/google.golang.org/grpc@v1.54.0/stream.go:760 +0x144
google.golang.org/grpc.newClientStreamWithParams({0x59c7c68, 0xc0135e2460}, 0x7bf00e0, 0xc001451180, {0x519c087, 0x1d}, {0x0, 0x0, 0x0, 0x0, ...}, ...)
	/go/pkg/mod/google.golang.org/grpc@v1.54.0/stream.go:345 +0xc76
google.golang.org/grpc.newClientStream.func2({0x59c7c68?, 0xc0135e2460?}, 0xc03baaace0?)
	/go/pkg/mod/google.golang.org/grpc@v1.54.0/stream.go:202 +0x9f
google.golang.org/grpc.newClientStream({0x59c7c68, 0xc0135e2460}, 0x7bf00e0, 0xc001451180, {0x519c087, 0x1d}, {0x0, 0x0, 0x0})
	/go/pkg/mod/google.golang.org/grpc@v1.54.0/stream.go:237 +0x674
google.golang.org/grpc.(*ClientConn).NewStream(0x18ae6c7?, {0x59c7c68?, 0xc0135e2460?}, 0xc03bb32c80?, {0x519c087?, 0x3998efd?}, {0x0?, 0x59c7c68?, 0xc0135e2460?})
	/go/pkg/mod/google.golang.org/grpc@v1.54.0/stream.go:162 +0x1ae
github.com/pingcap/kvproto/pkg/import_sstpb.(*importSSTClient).Write(0xf3457203a36cfba0?, {0x59c7c68?, 0xc0135e2460?}, {0x0?, 0x31?, 0x40?})
	/go/pkg/mod/github.com/pingcap/kvproto@v0.0.0-20230523065550-8b641fa69bf3/pkg/import_sstpb/import_sstpb.pb.go:2714 +0x6d
github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).writeToTiKV(0xc0014a6180, {0x59c7c68, 0xc0135e2460}, 0xc0310ccc00)
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/lightning/backend/local/region_job.go:231 +0xb5c
github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).executeJob(0xc0014a6180, {0x59c7c68, 0xc0135e2460}, 0xc0310ccc00)
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/lightning/backend/local/local.go:1357 +0xb4
github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).startWorker(0x4b1b6c0?, {0x59c7c68, 0xc0135e2460}, 0xc0006bc7e0, 0x0?, 0x0?)
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/lightning/backend/local/local.go:1271 +0x125
github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).doImport.func3()
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/lightning/backend/local/local.go:1547 +0x31
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/go/pkg/mod/golang.org/x/sync@v0.2.0/errgroup/errgroup.go:75 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
	/go/pkg/mod/golang.org/x/sync@v0.2.0/errgroup/errgroup.go:72 +0xa5
```

but in fact the TiKV has joined the cluster after injected 60 minutes failure.

[lightning-0-goroutine.txt](https://github.com/pingcap/tidb/files/12415237/lightning-0-goroutine.txt)

### LINKED ISSUE #48352: lightning may stuck at gRPC `RecvMsg`
## Bug Report

We found this goroutine stack

```
11 @ 0x1c2d6ee 0x1c3e185 0x21cf23c 0x225699a 0x2256987 0x2255c1f 0x22548da 0x2255a73 0x3d9bb96 0x3d999f1 0x3d98066 0x3d8942b 0x3d88f47 0x3d8be0e 0x3ac2a96 0x1c625c1
#	0x21cf23b	google.golang.org/grpc/internal/transport.(*Stream).waitOnHeader+0x7b			/go/pkg/mod/google.golang.org/grpc@v1.59.0/internal/transport/transport.go:327
#	0x2256999	google.golang.org/grpc/internal/transport.(*Stream).RecvCompress+0xb9			/go/pkg/mod/google.golang.org/grpc@v1.59.0/internal/transport/transport.go:342
#	0x2256986	google.golang.org/grpc.(*csAttempt).recvMsg+0xa6					/go/pkg/mod/google.golang.org/grpc@v1.59.0/stream.go:1070
#	0x2255c1e	google.golang.org/grpc.(*clientStream).RecvMsg.func1+0x1e				/go/pkg/mod/google.golang.org/grpc@v1.59.0/stream.go:927
#	0x22548d9	google.golang.org/grpc.(*clientStream).withRetry+0x139					/go/pkg/mod/google.golang.org/grpc@v1.59.0/stream.go:776
#	0x2255a72	google.golang.org/grpc.(*clientStream).RecvMsg+0x112					/go/pkg/mod/google.golang.org/grpc@v1.59.0/stream.go:926
#	0x3d9bb95	github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).doWrite.func5+0x2b5	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/lightning/backend/local/region_job.go:354
#	0x3d999f0	github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).doWrite+0x1890	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/lightning/backend/local/region_job.go:389
#	0x3d98065	github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).writeToTiKV+0x25	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/lightning/backend/local/region_job.go:189
#	0x3d8942a	github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).executeJob+0xaa	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/lightning/backend/local/local.go:1435
#	0x3d88f46	github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).startWorker+0x1c6	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/lightning/backend/local/local.go:1344
#	0x3d8be0d	github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).doImport.func5+0x2d	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/lightning/backend/local/local.go:1677
#	0x3ac2a95	golang.org/x/sync/errgroup.(*Group).Go.func1+0x55					/go/pkg/mod/golang.org/x/sync@v0.4.0/errgroup/errgroup.go:75
```

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close #48352 close #46321

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/local/region_job.go [hunk 1]
```diff
@@ -218,6 +218,14 @@ func (local *Backend) doWrite(ctx context.Context, j *regionJob) error {
 		failpoint.Return(err)
 	})
 
+	var cancel context.CancelFunc
+	// there are some strange blocking issues of gRPC like
+	// https://github.com/pingcap/tidb/issues/48352
+	// https://github.com/pingcap/tidb/issues/46321 and I don't know why 😭, so we
+	// set a timeout here to avoid blocking forever.
+	ctx, cancel = context.WithTimeout(ctx, 15*time.Minute)
```
[the remark this answers is not in the document]
Comment (lance6716) [reply br/pkg/lightning/backend/local/region_job.go:226]: No it should send about 96MB data to all peers, and from TiKV write RPC metrics it should be finished in a few seconds.

[the remark this answers is not in the document]
Comment (tangenta) [reply br/pkg/lightning/backend/local/region_job.go:226]: It is better to use [`context.WithCancelCause`](https://pkg.go.dev/context#WithCancelCause) and pass the reason to cancel function.

[the remark this answers is not in the document]
Comment (lance6716) [reply br/pkg/lightning/backend/local/region_job.go]: https://github.com/pingcap/tidb/pull/48355/commits/30a974306cfd28148f749cf46e3b4c4986861ba4

**Review by tangenta — APPROVED** [review]

**Review by ywqzzy — APPROVED** [review]

**Pull request merged**